- ## Vulnerability:
	    Insecure Data Handling
	    Unverified HTTPS Certificate

	## Location:
	```python
	response_google = urllib.request.urlopen(url)  # In handle_client function
	
	response_cloudflare = requests.get(url, params=params, headers=headers)  # In handle_client function
	
	response_alibaba = urllib.request.urlopen(url)  # In handle_client function
	
	response_quadnine = urllib.request.urlopen(url)  # In handle_client function
	
	response_quadnine_sec = urllib.request.urlopen(url)  # In handle_client function
	
	response_dns_ten = urllib.request.urlopen(url)  # In handle_client function
	
	response_dns_elev = urllib.request.urlopen(url)  # In handle_client function
	
	response_de_dus = urllib.request.urlopen(url)  # In handle_client function
	```

	## Hazard:
	    Using `urllib.request.urlopen` and `requests.get` without verifying the HTTPS certificate could expose the system to Man-in-the-Middle (MitM) attacks. Without HTTPS certificate validation, there's no assurance that the server being connected to is authentic. This leaves the communication channel vulnerable to interception, allowing attackers to eavesdrop, tamper with, or spoof data.

	## Fix Advice:
	1. **HTTPS Certificate Verification**: Utilize the SSL/TLS certificate validation mechanism to ensure the authenticity of the server. This involves verifying that the certificate presented by the server is signed by a trusted Certificate Authority (CA) and matches the expected hostname.
	    ```python
	    import ssl
	    
	    context = ssl.create_default_context()
	    context.check_hostname = True
	    context.verify_mode = ssl.CERT_REQUIRED
	    
	    response_google = urllib.request.urlopen(url, context=context)
	    ```
	    For `requests.get`, set `verify` parameter to `True`:
	    ```python
	    response_cloudflare = requests.get(url, params=params, headers=headers, verify=True)
	    ```

	2. **Exception Handling**: Properly handle exceptions to gracefully handle errors that may occur during HTTPS requests.
	    ```python
	    try:
	        response_google = urllib.request.urlopen(url, context=context)
	    except urllib.error.URLError as e:
	        print("Error occurred during Google DNS request:", e)
	        # Handle the error (e.g., log, fallback, etc.)
	    ```
	3. **Use of Known Certificate Authorities**: Explicitly specifying known CAs for certificate validation can enhance security.
	    ```python
	    context.load_verify_locations(cafile="/path/to/ca_certificate_bundle.pem")
	    ```

	## Additional Info:
	- Always validate HTTPS certificates to ensure secure communication.
	- Incorporate error handling to gracefully manage exceptions during HTTPS requests.
	- Regularly update the list of trusted Certificate Authorities to stay current with security standards.